### PR TITLE
Implement Digest trait for HashMatrix

### LINKED ISSUE #13: Implement `digest` traits
https://docs.rs/digest/0.9.0/digest/trait.Digest.html

### PR DESCRIPTION
**Pull request by basile-henry**
Closes #13

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -89,7 +89,9 @@ none of them fulfill these desiderata.
 #[macro_use]
 extern crate alloc;
 
-pub use crate::hash_matrix::{HashMatrix, constmatmul};
+use digest::{Digest, generic_array::typenum::Unsigned};
```
Comment (benwr) [inline src/lib.rs]: I think it's common practice to re-export `digest` and `Digest`.

File: src/lib.rs [hunk 2]
```diff
@@ -172,3 +174,42 @@ impl<T: BrombergHashable> BrombergHashable for alloc::rc::Rc<T> {
         (**self).bromberg_hash()
     }
 }
+
+impl Digest for HashMatrix {
```
Comment (benwr) [inline src/lib.rs]: I notice that the digest crate has some additional traits, which imply an instance of Digest. I haven't looked too far into it; any thoughts on whether it's better to implement those traits, vs implementing this one directly? If nothing else, I'd expect that implementing those would give a bit more granularity for downstream users.

**Review by benwr — COMMENTED** [review]
Thanks, this looks pretty awesome! Couple quick thoughts.

File: src/lib.rs (hunk 2 above)
Comment (basile-henry) [reply src/lib.rs]: You're right, I naively assumed they would automatically get implemented using this `Digest` trait, but it looks like it's the other way around: the `Digest` trait gets automatically implemented when the impl for the other traits exist.

I'll make that change tomorrow 😅

File: src/lib.rs (hunk 1 above)
Comment (basile-henry) [reply src/lib.rs]: Re-exported the traits, and also `GenericArray`.

File: src/hash_matrix.rs [hunk 3]
```diff
@@ -32,15 +33,53 @@ impl ToBigUint for U256 {
...
+}
+
+impl DigestString for HashMatrix {
     /// Produce a hex digest of the hash. This will be 128 hex digits.
     #[must_use]
     #[inline]
-    pub fn to_hex(self) -> String {
-        format!("{:016x}{:016x}{:016x}{:016x}",
+    fn to_hex(self) -> String {
+        format!("{:032x}{:032x}{:032x}{:032x}",
```
Comment (basile-henry) [inline src/hash_matrix.rs]: I fixed the string digest, it wouldn't always produce 128 hex digit long strings. :sweat_smile:

File: src/lib.rs [hunk 4]
```diff
@@ -89,7 +89,12 @@ none of them fulfill these desiderata.
 #[macro_use]
 extern crate alloc;
 
-pub use crate::hash_matrix::{HashMatrix, constmatmul};
+// Re-export digest traits
+pub use digest::{
```
Comment (benwr) [inline src/lib.rs:93]: I think adding `self` to this list will re-export `digest` itself, which is probably good to do.

**Review by benwr — APPROVED** [review]
Nice, thanks a ton for your work here!

**Comment by benwr** [comment]
I'll aim to cut a new version soon with these additions and fixes.

**Pull request merged**